File: android/src/main/kotlin/io/sentient/android/settings/tools/ToolsScreen.kt
// ---------------------------------------------------------------------------
// ToolsScreen — Tools settings page: per-MCP-group cards (master on/off Switch +
// expand to per-tool Allow/Ask/Deny/Off dropdowns), a "Gateway tools" card for
// native tools with no MCP group — most (skill tools) are per-person settable
// under the reserved transport-specific metadata, without a separate native bucket
// role-governed/read-only — and a Hermes built-ins card (per-toolset toggles).
// SLOW save (restart copy) via the shared SettingsEditChrome.
//
// The four-state permission semantics (see ToolsViewModel's file header) live in
// ToolsViewModel; this screen resolves them into RowSelect inputs, reading
// `state.pendingPermissions` (a patch overlay, NOT the loaded profile's own stored
// permissions) through the shared effectiveToolPermission helper. A group's tool
// list stays visible whenever expanded regardless of the master Switch's current
// reading — there is no "group is off, tools hidden" placeholder, since a tool's
// own row can independently read anything from Allow to Off. Copy mirrors webui
// tools-pane.
// ---------------------------------------------------------------------------
package io.sentient.android.settings.tools

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Switch
import androidx.compose.material3.SwitchDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.tooling.preview.Preview
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import io.sentient.android.settings.components.ApplyProgress
import io.sentient.android.settings.components.RowSelect
import io.sentient.android.settings.components.RowToggle
import io.sentient.android.settings.components.SettingsCard
import io.sentient.android.settings.components.SettingsEditScaffold
import io.sentient.android.settings.components.SettingsLoadStatus
import io.sentient.android.settings.components.SettingsPaneSub
import io.sentient.android.settings.components.SettingsTopBar
import io.sentient.android.theme.JetBrainsMono
import io.sentient.android.theme.LocalTokens
import io.sentient.android.theme.SentientTheme
import io.sentient.mobilesdk.design.Colors
import io.sentient.mobilesdk.settings.ProductToolGroupView
import io.sentient.mobilesdk.settings.McpCatalogView
import io.sentient.mobilesdk.settings.McpToolView
import io.sentient.mobilesdk.settings.ToolPermission
import io.sentient.mobilesdk.settings.ToolPermissionPatchMap
import io.sentient.mobilesdk.settings.effectiveToolPermission

private const val HEAD_SUB =
    "Tools available to the assistant each cycle. Toggle a group or built-in group on/off, or " +
        "expand to set individual tools to Allow, Ask, Deny or Off. Changes apply after Save — " +
        "the agent restarts to pick them up."

private const val NATIVE_TOOLS_SUB =
    "Built into the gateway itself, not an MCP group. Most rows (skill tools) are governed " +
        "per-person like any other tool; delegateTask is governed by role only — no stored key " +
        "can address it yet."

@Composable
fun ToolsScreen(
    vm: ToolsViewModel,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val state by vm.ui.collectAsStateWithLifecycle()
    SettingsEditScaffold(
        title = "Tools",
        screenTestTag = "settings-tools-screen",
        backTestTag = "settings-tools-back",
        saveTestTag = "settings-tools-save",
        dirty = state.dirty,
        apply = ApplyProgress(state.saving, state.restarting, state.alreadyApplying, state.applyError),
        onBack = onBack,
        onSave = vm::save,
        modifier = modifier,
    ) {
        ToolsBody(
            state = state,
            onToggleGroup = vm::toggleGroup,
            onSetToolPermission = vm::setToolPermission,
            onToggleToolset = vm::toggleToolset,
        )
    }
}

@Composable
private fun ColumnScope.ToolsBody(
    state: ToolsUiState,
    onToggleGroup: (String) -> Unit,
    onSetToolPermission: (String, String, ToolPermission) -> Unit,
    onToggleToolset: (String) -> Unit,
) {
    val catalog = state.catalog
    val original = state.original
    if (catalog == null || original == null) {
        SettingsLoadStatus(loading = state.loading, error = state.loadError)
        return
    }
    val enabled = !state.applyActive
    SettingsPaneSub(HEAD_SUB)
    val groupIds = remember(catalog) { catalog.groups.keys.sorted() }
    SettingsCard(title = "Product groups") {
        if (groupIds.isEmpty()) {
            EmptyRow("No tools configured. An admin can add them in gateway/config.yaml#mcp_catalog.")
        }
        groupIds.forEach { id ->
            val entry = catalog.groups[id] ?: return@forEach
            key(id) {
                ProductGroupSection(
                    id = id,
                    entry = entry,
                    permissions = state.pendingPermissions,
                    // Read through the SAME function the ViewModel's toggle inverts, so the
                    // Switch and the write can never disagree about which way a tap goes.
                    masterOn = isGroupOn(state.pendingPermissions, catalog, id),
                    controlsEnabled = enabled,
                    onToggleGroup = { onToggleGroup(id) },
                    onSetToolPermission = { toolName, permission -> onSetToolPermission(id, toolName, permission) },
                )
            }
        }
    }
    HermesBuiltinsCard(
        catalog = catalog,
        enabledToolsets = state.pendingToolsets ?: original.tools.toolsets ?: emptyList(),
        controlsEnabled = enabled,
        onToggleToolset = onToggleToolset,
    )
}

@Composable
private fun ProductGroupSection(
    id: String,
    entry: ProductToolGroupView,
    permissions: ToolPermissionPatchMap,
    masterOn: Boolean,
    controlsEnabled: Boolean,
    onToggleGroup: () -> Unit,
    onSetToolPermission: (String, ToolPermission) -> Unit,
) {
    var open by remember { mutableStateOf(false) }
    // Always "n/total tools", never a bare "off" — the count is a per-tool tally and the
    // Switch is the wildcard, and the two legitimately disagree between a master tap and
    // the save that resolves it. Collapsing the count to "off" made that disagreement read
    // as a contradiction. Mirrors webui tools-pane and the Hermes card below.
    val count = "${activeToolCount(permissions, id, entry)}/${entry.tools.size} tools"
    SectionHeaderRow(
        title = id,
        description = entry.description?.let { "$it · Default exposure: ${entry.defaultExposure.name.lowercase()}" }
            ?: "Default exposure: ${entry.defaultExposure.name.lowercase()}",
        countLabel = count,
        open = open,
        checked = masterOn,
        controlsEnabled = controlsEnabled,
        onExpand = { open = !open },
        onToggle = onToggleGroup,
        testTag = "settings-tools-group-$id",
    )
    if (open) {
        if (entry.tools.isEmpty()) {
            EmptyRow("No tools declared for this group.")
        } else {
            entry.tools.forEach { tool ->
                ToolPermissionRow(
                    tool = tool,
                    permission = effectiveToolPermission(permissions, id, tool),
                    controlsEnabled = controlsEnabled,
                    onSelect = { permission -> onSetToolPermission(tool.name, permission) },
                    testTag = "settings-tools-tool-$id-${tool.name}",
                )
            }
        }
    }
}

/**
 * One tool's four-state permission control: the tool name + description on the
 * left, a RowSelect (Allow/Ask/Deny/Off) on the right. Shared by
 * [ProductGroupSection]'s expanded per-tool list. [McpToolView.settable] is what
 * this row branches on to render read-only (a genuinely disabled,
 * non-interactive RowSelect — Compose's `clickable(enabled = false)` never
 * opens the menu) — never the tool's name.
 */
@Composable
private fun ToolPermissionRow(
    tool: McpToolView,
    permission: ToolPermission,
    controlsEnabled: Boolean,
    onSelect: (ToolPermission) -> Unit,
    testTag: String,
) {
    RowSelect(
        label = tool.name,
        sub = tool.description.ifBlank { null },
        options = toolPermissionSelectOptions,
        selectedValue = permission.wireValue,
        onSelect = { value -> toolPermissionFromWireValue(value)?.let(onSelect) },
        enabled = controlsEnabled && tool.settable,
        testTag = testTag,
    )
}

@Composable
private fun HermesBuiltinsCard(
    catalog: McpCatalogView,
    enabledToolsets: List<String>,
    controlsEnabled: Boolean,
    onToggleToolset: (String) -> Unit,
) {
    var open by remember { mutableStateOf(true) }
    val tools = remember(catalog) { catalog.hermesBuiltins.sortedWith(compareBy({ it.toolset }, { it.name })) }
    val active = tools.count { it.toolset in enabledToolsets }
    SettingsCard(title = "Hermes built-in tools") {
        SectionHeaderRow(
            title = "hermes",
            description = "Tools that run inside Hermes (memory, todo, web, browser…). Toggles operate on toolset groups.",
            countLabel = "$active/${tools.size} tools",
            open = open,
            checked = null,
            controlsEnabled = controlsEnabled,
            onExpand = { open = !open },
            onToggle = {},
            testTag = "settings-tools-hermes",
        )
        if (open) {
            tools.forEach { tool ->
                RowToggle(
                    label = tool.name,
                    sub = tool.description.ifBlank { tool.toolset },
                    checked = tool.toolset in enabledToolsets,
                    onCheckedChange = { onToggleToolset(tool.toolset) },
                    enabled = controlsEnabled,
                    testTag = "settings-tools-builtin-${tool.name}",
                )
            }
        }
    }
}

@Composable
private fun SectionHeaderRow(
    title: String,
    description: String?,
    countLabel: String,
    open: Boolean,
    checked: Boolean?,
    controlsEnabled: Boolean,
    onExpand: () -> Unit,
    onToggle: () -> Unit,
    testTag: String,
) {
    val tokens = LocalTokens.current
    Row(
        modifier = Modifier.fillMaxWidth().padding(horizontal = tokens.space.lg, vertical = tokens.space.sm).testTag(testTag),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(tokens.space.sm),
    ) {
        Row(
            modifier = Modifier.weight(1f).clickable(onClick = onExpand),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(tokens.space.sm),
        ) {
            Text(if (open) "⌄" else "›", color = Color(Colors.ink3), fontSize = tokens.type.base)
            Column(modifier = Modifier.weight(1f)) {
                Text(title, color = Color(Colors.ink), fontFamily = JetBrainsMono, fontSize = tokens.type.sm)
                if (description != null) {
                    Text(description, color = Color(Colors.ink3), fontSize = tokens.type.xs)
                }
            }
        }
        Text(countLabel, color = Color(Colors.ink3), fontSize = tokens.type.xs)
        if (checked != null) {
            Switch(
                checked = checked,
                onCheckedChange = { onToggle() },
                enabled = controlsEnabled,
                colors = SwitchDefaults.colors(
                    checkedThumbColor = Color(Colors.paper),
                    checkedTrackColor = Color(Colors.accent),
                    uncheckedTrackColor = Color(Colors.bgSunk),
                ),
            )
        }
    }
}

@Composable
private fun EmptyRow(text: String) {
    val tokens = LocalTokens.current
    Text(
        text = text,
        modifier = Modifier.fillMaxWidth().padding(horizontal = tokens.space.lg, vertical = tokens.space.sm),
        color = Color(Colors.ink3),
        fontSize = tokens.type.xs,
    )
}

@Preview
@Composable
private fun ToolsScreenPreview() {
    SentientTheme {
        Column {
            SettingsTopBar(title = "Tools", onBack = {})
            SettingsCard(title = "MCP groups") { EmptyRow("No tools configured.") }
        }
    }
}

File: ios/App/Settings/Tools/ToolsScreen.swift
// ---------------------------------------------------------------------------
// ToolsScreen — Soul-group "Tools" category page. A per-MCP-server card (master
// on/off toggle + expandable per-tool Allow/Ask/Deny/Off dropdowns) for each
// catalog server, a "Gateway tools" card for native tools with no MCP server —
// most (skill tools) are per-person settable under the reserved "native"
// permission namespace, delegateTask stays role-governed/read-only — and the
// Hermes built-in toolsets card. SLOW save (PUT profile → apply-with-restart).
//
// The permission-resolution semantics live in ToolsViewModel (pinned to the
// webui tools-pane); this view resolves them into stateless ToolsServerCard /
// ToolPermissionRow inputs. The shared apply bar receives the existing dirty,
// save, and native discard-confirmation actions.
// ---------------------------------------------------------------------------
import SwiftUI
import MobileData

struct ToolsScreen: View {
    let settings: SettingsComponent
    let onBack: () -> Void

    @State private var vm: ToolsViewModel
    @State private var openServers: Set<String> = []
    @State private var builtInsOpen = true
    @State private var permissionGuideOpen = false
    @State private var showDiscard = false

    init(settings: SettingsComponent, onBack: @escaping () -> Void) {
        self.settings = settings
        self.onBack = onBack
        _vm = State(initialValue: ToolsViewModel(settings: settings))
    }

    var body: some View {
        SettingsPageScaffold(
            title: "Tools", screenId: "settings-tools-screen",
            onBack: attemptBack, allowsInteractiveBack: !vm.isDirty,
            backAccessibilityId: "settings-tools-back"
        ) {
            switch vm.phase {
            case .loading:
                SoulLoadingRow()
            case .failed(let message):
                AsyncNotice(kind: .error, title: "Couldn't load capabilities", detail: message) {
                    Task { await vm.load() }
                }
            case .ready:
                capabilityOverview
                groupsSection
                builtInCard
            }
        }
        .designApplyBarDock(
            isDirty: vm.isDirty,
            state: applyState,
            discardAccessibilityId: "settings-tools-discard",
            applyAccessibilityId: "settings-tools-save",
            onDiscard: attemptBack,
            onApply: { Task { await vm.save() } }
        )
        .task { await vm.load() }
        .confirmationDialog("Discard changes?", isPresented: $showDiscard, titleVisibility: .visible) {
            Button("Discard", role: .destructive) { onBack() }
            Button("Keep editing", role: .cancel) {}
        }
    }

    private var applyState: DesignApplyState {
        switch vm.save {
        case .idle: .idle
        case .saving: .saving
        case .restarting: .restarting
        case .alreadyApplying: .alreadyApplying
        case .applied: .applied
        case .failed(let message): .failed(message)
        }
    }

    private var capabilityOverview: some View {
        VStack(alignment: .leading, spacing: Space.md) {
            Text("Control what Sentient can use")
                .designText(.label)
                .fontWeight(.semibold)
                .foregroundStyle(DuskColors.ink)
                .accessibilityAddTraits(.isHeader)
            Text("Role and resource access checks still apply to every request.")
                .designText(.supporting)
                .foregroundStyle(DuskColors.ink2)
            DesignDisclosureGroup(isExpanded: permissionGuideOpen) {
                DesignDisclosureButton(
                    isExpanded: permissionGuideOpen,
                    accessibilityLabel: "Permission meanings: Allow, Ask, Deny, Off",
                    accessibilityId: "settings-tools-permission-guide",
                    action: { permissionGuideOpen.toggle() }
                ) {
                    VStack(alignment: .leading, spacing: Space.xs) {
                        Text("Permission meanings")
                            .designText(.label)
                            .foregroundStyle(DuskColors.ink)
                        Text(ToolPermission.selectOptions.map(\.label).joined(separator: " · "))
                            .designText(.supporting)
                            .foregroundStyle(DuskColors.ink2)
                    }
                }
            } content: {
                VStack(alignment: .leading, spacing: Space.sm) {
                    ForEach(ToolPermission.selectOptions, id: \.id) { option in
                        if let permission = ToolPermission(wireValue: option.id) {
                            Text("\(option.label) · \(permission.meaning)")
                                .designText(.supporting)
                                .foregroundStyle(DuskColors.ink2)
                        }
                    }
                }
            }
        }
    }

    @ViewBuilder
    private var groupsSection: some View {
        if vm.groupIds.isEmpty {
            AsyncNotice(
                kind: .empty,
                title: "No connected capabilities",
                detail: "Built-in capabilities are still available below."
            )
            .accessibilityIdentifier("settings-tools-empty")
        } else {
            ForEach(vm.groupIds, id: \.self) { id in
                if let entry = vm.catalog?.groups[id] {
                    groupCard(id, entry)
                }
            }
        }
    }

    private func groupCard(_ id: String, _ entry: ProductToolGroupView) -> some View {
        let rows = entry.tools.map { tool in
            ToolPermissionRowModel(
                id: tool.name,
                name: tool.name,
                description: tool.description,
                permission: vm.toolPermission(id, tool),
                settable: tool.settable
            )
        }
        return ToolsServerCard(
            id: id,
            serverDescription: entry.description_,
            masterOn: vm.isGroupMasterOn(id, entry),
            isOpen: openServers.contains(id),
            toolRows: rows,
            onToggleOpen: { toggleOpen(id) },
            onToggleServer: { turnOn in vm.setGroupMaster(id, entry.tools.map { $0.name }, turnOn) },
            onToolChange: { toolName, permission in vm.setToolPermission(id, toolName, permission) }
        )
    }

    @ViewBuilder
    private var builtInCard: some View {
        let builtins = (vm.catalog?.hermesBuiltins ?? []).sorted {
            $0.toolset == $1.toolset ? $0.name < $1.name : $0.toolset < $1.toolset
        }
        DesignCard {
            DesignDisclosureGroup(isExpanded: builtInsOpen) {
                DesignDisclosureButton(
                    isExpanded: builtInsOpen,
                    accessibilityLabel: builtInsOpen ? "Collapse built-in capabilities" : "Expand built-in capabilities",
                    accessibilityId: "settings-tools-hermes-expand",
                    action: { builtInsOpen.toggle() }
                ) {
                    ViewThatFits(in: .horizontal) {
                        HStack(spacing: Space.sm) {
                            builtInHeading
                            Spacer(minLength: Space.sm)
                            enabledCount(vm.hermesActiveCount(builtins), total: builtins.count)
                        }
                        VStack(alignment: .leading, spacing: Space.xs) {
                            builtInHeading
                            enabledCount(vm.hermesActiveCount(builtins), total: builtins.count)
                        }
                    }
                }
            } content: {
                ForEach(builtins, id: \.name) { tool in
                    DesignToggleRow(
                        title: capabilityName(tool.name),
                        detail: builtInDetail(tool.description, toolset: tool.toolset),
                        isOn: Binding(
                            get: { vm.isToolsetOn(tool.toolset) },
                            set: { _ in vm.toggleToolset(tool.toolset) }
                        ),
                        accessibilityId: "settings-tools-builtin-\(tool.name)"
                    )
                }
            }
        }
    }

    private var builtInHeading: some View {
        VStack(alignment: .leading, spacing: Space.xs) {
            Text("Built-in capabilities")
                .designText(.label)
                .fontWeight(.semibold)
                .foregroundStyle(DuskColors.ink)
            Text("Related capabilities may be enabled or disabled together.")
                .designText(.supporting)
                .foregroundStyle(DuskColors.ink2)
        }
    }

    private func enabledCount(_ count: Int, total: Int) -> some View {
        Text("\(count) of \(total) enabled")
            .designText(.caption)
            .foregroundStyle(DuskColors.ink2)
    }

    private func builtInDetail(_ description: String, toolset: String) -> String {
        let group = capabilityName(toolset)
        return description.isEmpty ? group : "\(description) · \(group)"
    }

    private func toggleOpen(_ id: String) {
        if openServers.contains(id) { openServers.remove(id) } else { openServers.insert(id) }
    }

    private func attemptBack() {
        if vm.isDirty { showDiscard = true } else { onBack() }
    }
}

#Preview("servers") {
    NavigationStack {
        SettingsPageScaffold(title: "Tools", screenId: "settings-tools-screen") {
            ToolsServerCard(
                id: "home-assistant", serverDescription: "Smart-home control.",
                masterOn: true, isOpen: true,
                toolRows: [
                    ToolPermissionRowModel(
                        id: "1", name: "turn_on", description: "Turn a device on.", permission: .allow, settable: true
                    ),
                    ToolPermissionRowModel(
                        id: "2", name: "turn_off", description: "Turn a device off.", permission: .off, settable: true
                    ),
                ],
                onToggleOpen: {}, onToggleServer: { _ in }, onToolChange: { _, _ in }
            )
        }
    }
    .preferredColorScheme(.dark)
}
